### PR TITLE
Include arguments in function suggestions

### PR DESCRIPTION
**Pull request by koljonen**
## Description
E.g. instead of suggesting `my_func()`, suggest `my_func(your_arg:=)` or `my_func(your_arg text)`, depending on the context. There are config settings in the form of template strings (e.g. `{arg_name} := {arg_default}`) for customizing the formatting of the arguments and a setting for at what number of args to switch from oneliners to multiline arg lists.

## Checklist
- [x] I've added this contribution to the `changelog.md`.
- [x] I've added my name to the `AUTHORS` file (or it's already there).

### CODE REVIEW

**Comment by darikg** [comment]
I'm sorry, I haven't been intentionally ignoring this! Just super swamped the past couple weeks. I'll try to have a look this weekend

File: pgcli/pgclirc [hunk 1]
```diff
@@ -71,6 +71,18 @@ qualify_columns = if_more_than_one_table
...
+# {max_arg_len} is available for for alignment purposes (for multiline arg lists).
+# (See https://docs.python.org/2/library/string.html#formatstrings for details.)
+call_arg_style = {arg_name: <{max_arg_len}} := {arg_default}
+call_arg_display_style = {arg_name}
+# Func calls with > oneliner_max are prefixed with \n\t
+call_arg_oneliner_max = 2
+# Formatting of function arguments in function signatures, e.g. after DROP FUNCTION.
+# Supported tags: {arg_name}, {arg_num}, {arg_type} and {arg_default}
+signature_arg_style = {arg_name} {arg_type}
+
```
Comment (amjith) [inline pgcli/pgclirc]: Can we choose reasonable defaults and remove these config options? 

I'm not a fan of adding more configuration.

**Comment by darikg** [comment]
Just starting to look at this now. Did you consider instead suggesting argument names as a distinct suggestion type? E.g. `select my_func(` and `select my_func(y` would suggest `your_arg:=`? That might sidestep having to tailor pgclirc to the perfect formatting strings

**Comment by koljonen** [comment]
I did consider it briefly, but it seemed more complicated to code and like more typing/thinking to use, especially with overloaded functions. I actually started working on this for the `DROP FUNCTION` case (where suggesting args one at a time would not be very useful) but realized I could easily expand the functionality to function calls.

File: pgcli/pgclirc (hunk 1 above)
Comment (koljonen) [inline pgcli/pgclirc]: Sure. As long as I can pass in weird templates when using the completer outside pgcli, I'm happy. (A couple of coworkers are working on an Atom plugin that uses the pgcli completer.)

File: pgcli/packages/sqlcompletion.py [hunk 2]
```diff
@@ -427,8 +430,17 @@ def suggest_based_on_last_token(token, stmt):
 
         return tuple(suggest)
 
-    elif token_v in ('table', 'view', 'function'):
-        # E.g. 'DROP FUNCTION <funcname>', 'ALTER TABLE <tablname>'
+    elif token_v == 'function':
+        schema = stmt.get_identifier_schema()
+        if stmt.get_previous_token(token).value.lower() == 'drop':
```
Comment (darikg) [inline pgcli/packages/sqlcompletion.py]: the `if` and `elif` can be combined here, right? 
`if (...).value.lower() in ('drop', 'alter'):`

File: pgcli/pgcompleter.py [hunk 3]
```diff
@@ -600,23 +633,76 @@ def get_from_clause_item_matches(self, suggestion, word_before_cursor):
...
+        else:
+            return '(' + ', '.join(a for a in args if a) + ')'
+
+    def _format_arg(self, template, arg, arg_num, max_arg_len):
+        if not template:
+            return None
+        if arg.has_default:
+            arg_default = 'NULL' if arg.default is None else arg.default
+            # Remove trailing ::(schema.)type
+            arg_default = re.sub(r'::[\w\.]+(\[\])?$', '', arg_default)
```
Comment (darikg) [inline pgcli/pgcompleter.py]: The regex could be made a `re.compile`d module level global so it only needs to be compiled once

File: pgcli/pgcompleter.py [hunk 4]
```diff
@@ -194,6 +209,19 @@ def extend_functions(self, func_data):
 
             self.all_completions.add(func)
 
+        self._refresh_suffix_cache()
+
+    def _refresh_suffix_cache(self):
+        self._suffix_cache = {
```
Comment (darikg) [inline pgcli/pgcompleter.py]: Can you add a comment explaining the suffix cache here?

File: pgcli/packages/parseutils/meta.py [hunk 5]
```diff
@@ -42,19 +83,45 @@ def __eq__(self, other):
     def __ne__(self, other):
         return not self.__eq__(other)
 
+    def _sign(self):
```
Comment (darikg) [inline pgcli/packages/parseutils/meta.py]: why "sign"?

**Comment by darikg** [comment]
Oh yeah, 'drop function' is a perfect case where it's better to suggest the arguments all at once. For calling functions i'm not so sure. I don't have a good feel for when most people use named arguments or not, or if they do so consistently.

**Comment by j-bennet** [comment]
@koljonen What is the status of this PR? Are you still working on it? It would be great to have this functionality in the new release.

**Comment by koljonen** [comment]
Sorry about the delay; been really busy and didn't notice the comments until now. I'm having a look now.

File: pgcli/pgcompleter.py (hunk 4 above)
Comment (koljonen) [inline pgcli/pgcompleter.py]: I've also renamed it arg_list_cache to make things clearer.

File: pgcli/packages/parseutils/meta.py (hunk 5 above)
Comment (koljonen) [inline pgcli/packages/parseutils/meta.py]: I'm renaming it to _signature, for clarity.

**Comment by koljonen** [comment]
@darikg I'll do a squash and rebase once you've had a look.

File: pgcli/completion_refresher.py [hunk 6]
```diff
@@ -140,3 +136,7 @@ def refresh_casing(completer, executor):
     if os.path.isfile(casing_file):
         with open(casing_file, 'r') as f:
             completer.extend_casing([line.strip() for line in f])
+
+@refresher('functions')
```
Comment (bersace) [inline pgcli/completion_refresher.py]: @koljonen looks like you just moved the function definition. Can you explain me this ?

File: pgcli/pgcompleter.py [hunk 7]
```diff
@@ -29,15 +29,25 @@
...
-def SchemaObject(name, schema=None, function=False):
-    return _SchemaObject(name, schema, function)
+
+_SchemaObject = namedtuple('SchemaObject', 'name schema meta')
+def SchemaObject(name, schema=None, meta=None):
+    return _SchemaObject(name, schema, meta)
 
 _Candidate = namedtuple(
-    'Candidate', ['completion', 'prio', 'meta', 'synonyms', 'prio2']
+    'Candidate', 'completion prio meta synonyms prio2 display'
```
Comment (bersace) [inline pgcli/pgcompleter.py]: @koljonen Can you explain me the rationale behind this change ?

Comment (koljonen) [inline pgcli/pgcompleter.py]: I suppose you mean the style change? I probably just thought it looked cleaner and would keep us from running into PEP8's max line length.

File: pgcli/completion_refresher.py (hunk 6 above)
Comment (koljonen) [inline pgcli/completion_refresher.py]: `PGCompleter.extend_functions()` uses the casing info (in `_format_arg()`) added by `PGCompleter.extend_casing()` (and the refresher functions are called in the order they're defined) so the casing and function refreshers need to be in that order.

Comment (bersace) [inline pgcli/completion_refresher.py]: Ok, I missed this !! Thanks for the explanation :) 👍

**Comment by koljonen** [comment]
I've now squashed and rebased, fixed an obscure bug (the completer crashed if one tried to use a column named `function`), made the Travis CI linter happy and added tests.
Also, I might add I've been using this at work for a couple of months now, so it's gotten a fair bit of manual testing.

**Comment by amjith** [comment]
@darikg Are you satisfied with the changes pushed by the recent commits? Can we merge this PR?

**Comment by amjith** [comment]
@koljonen I'll take over the review from @darikg. I think he might be busy. 

Can you give me an example query to trigger this feature?

**Comment by koljonen** [comment]
Assuming you have a function like `create function test_func(_arg1 int, _arg2 text default 'foo', _arg3 numeric default 42) returns int language sql as $$ select 23$$`, trigger completions for e.g. `select test_func` or `drop function test_func`.

**Comment by amjith** [comment]
Does this not work with built-in functions? Say `sum()` or `count()` etc.

**Comment by koljonen** [comment]
`pg_proc.proargnames` appears to be NULL for builtins.

**Comment by amjith** [comment]
Works beautifully. 👍 

1__0_1__tmux__-__pgcli___users_amjith_code_python_pgcli____tmux_


/cc @qdw

**Pull request merged**